### PR TITLE
Add `__all__` to modules beginning with 'm'

### PR DESCRIPTION
**Pull request by AlexWaygood**
Also:

- Improve `multiprocessing.connection._ConnectionBase.__init__`. (`SupportsIndex` can be used unconditionally if we import it from `typing_extensions` instead of `typing`).
- Add `multiprocessing.shared_memory` to `VERSIONS`; remove the module-level `sys.version_info` guard and the allowlist entries.

### CODE REVIEW

File: stdlib/multiprocessing/__init__.pyi [hunk 1]
```diff
@@ -29,6 +29,33 @@ if sys.version_info >= (3, 8):
 if sys.platform != "win32":
     from multiprocessing.context import ForkContext, ForkServerContext
 
+# fmt: off
```
Comment (JelleZijlstra) [inline stdlib/multiprocessing/__init__.pyi:32]: I'd rather we just let Black format this.

Comment (AlexWaygood) [reply stdlib/multiprocessing/__init__.pyi:32]: Okay!

Comment (JelleZijlstra) [reply stdlib/multiprocessing/__init__.pyi]: Thanks! (pet peeve: it's "free rein", as in letting go of your horse's reins)

Comment (AlexWaygood) [reply stdlib/multiprocessing/__init__.pyi]: Lol, of course. Too much writing of code recently, not enough writing of words...

**Pull request merged**